src/lib/reasoners/satml_frontend.ml

```diff
@@ -1270,13 +1270,20 @@ module Make (Th : Theory.S) : Sat_solver_sig.S = struct
 
   let push env to_push =
     Util.loop ~f:(fun _n () acc ->
-        let expr_guard, atom_guard = create_guard acc in
-        SAT.push acc.satml atom_guard;
-        Stack.push expr_guard acc.guards.stack_guard;
-        Steps.push_steps ();
-        {acc with guards =
-                    {acc.guards with
-                     current_guard = expr_guard;} }
+        try
+          let expr_guard, atom_guard = create_guard acc in
+          SAT.push acc.satml atom_guard;
+          Stack.push expr_guard acc.guards.stack_guard;
+          Steps.push_steps ();
+          {acc with guards =
+                      {acc.guards with
+                       current_guard = expr_guard;} }
+        with
+        | Util.Step_limit_reached _ ->
+          (* If we reached the step limit while pushing, we have to cancel the
+             push.
+             TODO: properly cancel the push. *)
+          env
```

Comment:
Returning `env` here is not correct for satML. Because the underlying satml solver uses mutability, it is shared between `env` and `acc`; if we reach the steps limit here, `new_vars` (called through `create_guard`) can leave the satml solver in an inconsistent state with variables partially added. Plus, `env` will be out of sync with the content of the satml solver and weird things will follow.

I see two solutions here — either we make sure that `new_vars` returns a valid environment here, so that `acc` is correct in the exception handler and we can then pop a corresponding amount, or we temporarily disable the step limit while pushing. I'd prefer the second option (even though it can cause going slightly over the step limit) because the first one is opening up a can of worms where we *will* mess up and introduce subtle bugs in the future.

Reply:
> Returning env here is not correct for satML. 

That's the point of the `TODO`, we need to actually cancel the push instead of just returning the env. But disabling the step limit is a better idea, I'll go for it.

Reply:
What I mean by "not correct" here is worse — it will cause the type of crashes that we see in #946 , see my toplevel comment.